### PR TITLE
Add AppNexus adapter for AMP

### PR DESCRIPTION
**Pull request by mkendall07**

### CODE REVIEW

File: ads/_config.js [hunk 1]
```diff
@@ -117,6 +118,11 @@ export const adPreconnect = {
   webediads: [
     'https://eu1.wbdds.com',
   ],
+  appnexus: [
+    'https://ib.adnxs.com',
+    'https://secure.adnxs.com',
+    'https://acdn.adnxs.com/',
```
Comment (cramforce) [inline ads/_config.js]: No need to repeat host used in prefetch above.

File: ads/appnexus.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+const APPNEXUS_AST_URL = 'https://acdn.adnxs.com/ast/ast.js';
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function appnexus(global, data) {
+  // in case we pass the ttj url to use, simply call it and return
+  if (data.src) {
+    validateSrcPrefix('https:', data.src);
```
Comment (cramforce) [inline ads/appnexus.js]: Are you sure you need this? We do allow this type of "XSS", but it makes your component unavailable to publishers who use custom ad iframes.

File: ads/appnexus.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+    writeScript(global, data.src);
+    return;
+  }
+  // otherwise, use json configuration to load ast
+  if (context.isMaster) { // in case we are in the master iframe, we load AST
+    apntag = (typeof apntag !== 'undefined') ? apntag : {};
+    apntag.anq = apntag.anq || [];
+
+    writeScript(global, APPNEXUS_AST_URL);
```
Comment (cramforce) [inline ads/appnexus.js]: Can this be async?

File: ads/appnexus.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      apntag.anq.push(function() {
+        //output console information
+        apntag.debug = data.debug || false;
+        //set global page options
+        apntag.setPageOpts(data.pageOpts);
+      });
+    }
+
+    for (let i = 0; i < data.adUnits.length; ++i) {
+      (function(i) {
```
Comment (cramforce) [inline ads/appnexus.js]: No need with `let`

File: ads/appnexus.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        });
+      })(i);
+    }
+
+    apntag.anq.push(function() {
+      apntag.loadTags();
+    });
+  }
+  // then for all ad units, define the ad placement and show the ad
+  global.docEndCallback = function() {
```
Comment (cramforce) [inline ads/appnexus.js]: The tag appears to support async calls. Do you need to use this sync mechanism?

File: ads/appnexus.js (hunk 2 above)
Comment (mkendall07) [inline ads/appnexus.js]: @cramforce 
Can you explain this part a bit? Sorry not following.

Comment (cramforce) [inline ads/appnexus.js]: Basically here one can control the `data.src` via the URL, thus run arbitrary JS. That is fine on our sandbox frames, but for pubs who host ads on their own origin, this is an XSS against their origin. We thus cannot allow ad types that have such issues for pubs that use their own iframes.

File: ads/appnexus.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+ * @param {!Object} data
+ */
+export function appnexus(global, data) {
+  let useTtj = false;
+  const args = [];
+  args.push('size=' + data.width + 'x' + data.height);
+  if (data.tagid) {
+    args.push('id=' + encodeURIComponent(data.tagid));
+    writeScript(global, constructTtj(args));
+    usettj = true;
```
Comment (protonate) [inline ads/appnexus.js]: This can be removed as a leaked global var -- `usettj` instead of `useTtj` -- and the var gets set in `construct Ttj()`

**Comment by mkendall07** [comment]
@cramforce review notes have been addressed with a refactor. Please review when you can. Thanks

File: ads/appnexus.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    args.push('member=' + encodeURIComponent(data.member));
+    args.push('inv_code=' + encodeURIComponent(data.code));
+    writeScript(global, constructTtj(args));
+    return;
+  }
+
+  function constructTtj(args) {
+    let url = 'https://ib.adnxs.com/ttj?';
+    for (i = 0; i < args.length; i++) {
+      url += args[i] + '&';
```
Comment (cramforce) [inline ads/appnexus.js]: Please use `encodeURIComponent`

File: ads/appnexus.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+  if (!apntag) {
+    apntag = context.master.apntag;
+
+    //preserve a global reference
+    global.apntag = context.master.apntag;
+  }
+
+  apntag.anq.push(() => {
+
```
Comment (cramforce) [inline ads/appnexus.js]: nit: remove empty line.

File: ads/appnexus.js (hunk 7 above)
Comment (mkendall07) [inline ads/appnexus.js]: @cramforce 
They are already encoded above (line 33,38,39). I don't want to encode the `=`.

Comment (cramforce) [inline ads/appnexus.js]: I see. Please add a comment here.

File: ads/appnexus.js (hunk 8 above)
Comment (mkendall07) [inline ads/appnexus.js]: removed.

File: ads/appnexus.js (hunk 7 above)
Comment (mkendall07) [inline ads/appnexus.js]: Done

**Comment by cramforce** [comment]
LGTM

**Pull request merged**